flow_runner.go

```diff
@@ -142,37 +148,122 @@ func (f *flowRunner) runTask(ctx context.Context, task Task) bool {
+		paramValues[f.verbose.Name()] = f.paramValues[f.verbose.Name()]
 	}
 
-	// report task start
-	fmt.Fprintf(w, "===== TASK  %s\n", task.Name)
+	failed := false
+	measuredCommand := func(tf *TF) {
+		w := tf.Output()
+		if (f.verbose != nil) && !f.verbose.Get(tf) {
+			w = &strings.Builder{}
+		}
 
-	// run task
-	runner := Runner{
-		Ctx:      ctx,
-		TaskName: task.Name,
-		Verbose:  f.verbose,
-		Params:   f.params,
-		Output:   w,
+		// report task start
+		fmt.Fprintf(w, "===== TASK  %s\n", tf.Name())
+
+		// run task
+		runner := Runner{
+			Ctx:         tf.Context(),
+			TaskName:    tf.Name(),
+			ParamValues: tf.paramValues,
+			Output:      w,
+		}
+		result := runner.Run(task.Command)
+
+		// report task end
+		status := "PASS"
+		switch {
+		case result.Failed():
+			status = "FAIL"
+			failed = true
+		case result.Skipped():
+			status = "SKIP"
+		}
+		fmt.Fprintf(w, "----- %s: %s (%.2fs)\n", status, tf.Name(), result.Duration().Seconds())
+
+		if sb, ok := w.(*strings.Builder); ok && result.failed {
+			io.Copy(tf.Output(), strings.NewReader(sb.String())) //nolint // not checking errors when writting to output
+		}
+	}
+
+	measuredRunner := Runner{
+		Ctx:         ctx,
+		TaskName:    task.Name,
+		ParamValues: paramValues,
+		Output:      f.output,
+	}
+	measuredRunner.Run(measuredCommand)
+
+	return !failed
+}
+
+func (f *flowRunner) unusedParams() []string {
+	remainingParams := make(map[string]struct{})
+	for key := range f.params {
+		remainingParams[key] = struct{}{}
 	}
-	result := runner.Run(task.Command)
+	if f.verbose != nil {
+		delete(remainingParams, f.verbose.Name())
+	}
+	for _, task := range f.tasks {
+		for _, param := range task.Parameters {
+			delete(remainingParams, param.Name())
+		}
+	}
+	unusedParams := make([]string, 0, len(remainingParams))
+	for key := range remainingParams {
+		unusedParams = append(unusedParams, key)
+	}
+	return unusedParams
+}
 
-	// report task end
-	status := "PASS"
-	switch {
-	case result.Failed():
-		status = "FAIL"
-	case result.Skipped():
-		status = "SKIP"
+func printUsage(f *flowRunner) {
+	fmt.Fprintf(f.output, "Usage: [flag(s)] task(s)\n")
+	fmt.Fprintf(f.output, "Flags:\n")
+	w := tabwriter.NewWriter(f.output, 1, 1, 4, ' ', 0)
+	keys := make([]string, 0, len(f.params))
+	for key := range f.params {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		param := f.params[key]
+		fmt.Fprintf(w, "  %s\t%s\tDefault: %s\t%s\n",
+			param.info.shortFlag(), param.info.longFlag(), param.newValue().String(), param.info.Usage)
 	}
-	fmt.Fprintf(w, "----- %s: %s (%.2fs)\n", status, task.Name, result.Duration().Seconds())
+	w.Flush() //nolint // not checking errors when writing to output
 
-	if sb, ok := w.(*strings.Builder); ok && result.failed {
-		io.Copy(f.output, strings.NewReader(sb.String())) //nolint // not checking errors when writting to output
+	fmt.Fprintf(f.output, "Tasks:\n")
+	w = tabwriter.NewWriter(f.output, 1, 1, 4, ' ', 0)
```

Comment:
I am not sure if this line is needed

Reply:
Thank you, I learned new behaviour of the writer.